libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+            if let Ok(mut remote_connections) = remote_connections.try_upgrade() {
+                for failed_to_send in failed_to_send {
+                    remote_connections.remove(&failed_to_send);
+                }
+            }
+        }
+
+        debug!(target: "libp2p-floodsub", "Message queued for {} remotes", num_dispatched);
+    }
+}
+
+/// Implementation of `Stream` that provides messages for the subscribed topics you subscribed to.
+pub struct FloodSubReceiver {
+    inner: mpsc::UnboundedReceiver<Message>,
+}
+
+impl Stream for FloodSubReceiver {
+    type Item = Message;
+    type Error = IoError;
+
+    #[inline]
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        self.inner
+            .poll()
+            .map_err(|_| unreachable!("UnboundedReceiver cannot err"))
+    }
+}
+
+impl fmt::Debug for FloodSubReceiver {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("FloodSubReceiver").finish()
+    }
+}
+
+/// A message received by the floodsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Message {
+    /// Remote that sent the message.
+    pub source: Multiaddr,
+
+    /// Content of the message. Its meaning is out of scope of this library.
+    pub data: Vec<u8>,
+
+    /// List of topics of this message.
+    ///
+    /// Each message can belong to multiple topics at once.
+    pub topics: Vec<TopicHash>,
+}
+
+/// Implementation of `Future` that must be driven to completion in order for floodsub to work.
+pub struct FloodSubFuture {
+    inner: Box<Future<Item = (), Error = IoError>>,
+}
+
+impl Future for FloodSubFuture {
+    type Item = ();
+    type Error = IoError;
+
+    #[inline]
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        self.inner.poll()
+    }
+}
+
+impl fmt::Debug for FloodSubFuture {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("FloodSubFuture").finish()
+    }
+}
+
+// Handles when a packet is received on a connection.
+//
+// - `bytes` contains the raw data.
+// - `remote_addr` is the address of the sender.
+fn handle_packet_received(
+    bytes: BytesMut,
+    inner: Arc<Inner>,
+    remote_addr: &Multiaddr,
+) -> Result<(), IoError> {
+    trace!(target: "libp2p-floodsub", "Received packet from remote");
+
+    // Parsing attempt.
+    let mut input = match protobuf::parse_from_bytes::<rpc_proto::RPC>(&bytes) {
+        Ok(msg) => msg,
+        Err(err) => {
+            debug!(target: "libp2p-floodsub", "Failed to parse protobuf message ; err = {:?}", err);
+            return Err(err.into());
+        }
+    };
+
+    // Update the topics the remote is subscribed to.
+    if !input.get_subscriptions().is_empty() {
+        let remote_connec = inner.remote_connections.write();
+        // TODO: what if multiple entries?
+        let remote = remote_connec.get(remote_addr).unwrap();
```

Comment:
`expect` with a message describing what you're expecting and why. Even indexing would be better because at least then there's a file and location.